html/user/openssl_sign_action.php

```diff
@@ -0,0 +1,134 @@
+// http://boinc.berkeley.edu
+// Copyright (C) 2008 University of California
+//
+// BOINC is free software; you can redistribute it and/or modify it
+// under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation,
+// either version 3 of the License, or (at your option) any later version.
+//
+// BOINC is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+// See the GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
+
+require_once("../inc/boinc_db.inc");
+require_once("../inc/user.inc");
+require_once("../inc/util.inc");
+
+// Needs to be defined, unknown use.
+check_get_args(array("tnow", "ttok"));
+
+// Check the user is online
+$user = get_logged_in_user();
+check_tokens($user->authenticator);
+
+// For some reason, I'm only able to use numbers, no letters.
+// Need to provide security, but full functionality is likewise desirable.
+$user_data = sanitize_tags(post_str("user_data", true));
+
+// Following line a potential bug.
+// $user_data = BoincDb::escape_string($user_data);
+
+//Has the user entered data with length > 0 & <= 256?
+if (($user_data) && (strlen($user_data) <= 256)) {
+	// Retrieve the logged in user's UserID.
+	$userid = $user->id;
+
+	// Create the message which will be signed.
+	$message_data = "$userid $user_data";
+
+	// Create a digest of the message_data.
+	// Without the digest, the openssl_sign function fails to sign lower case characters.
+	// We can prove the integrity of the message with a digest.
+	$message_digest = openssl_digest ($message_data , "sha512"); 
+
+	// How to generate required keys in /project/keys/ folder:
+	// openssl genpkey -algorithm RSA -out external_system_key_private.pem -pkeyopt rsa_keygen_bits:2048
+	// openssl rsa -pubout -in external_system_key_private.pem -out external_system_key_public.pem
+
+	$private_key_path = "file:///root/project/keys/external_system_key_private.pem";
+	$public_key_path = "file:///root/project/keys/external_system_key_public.pem";
+
+	// Check that the private key file exists where specified. If not, redirect to error page.
+	if (!file_exists($private_key_path)) {
+		error_page(tra("The private key does not exist. Proof of UserID currently not possible."));
+	}
+
+	// Check that the public key file exists where specified. If not, redirect to error page.
+	if (!file_exists($public_key_path)) {
+		error_page(tra("The public key does not exist. Proof of UserID currently not possible."));
+	}
+
+	// Fetch private key from file
+	$private_key_pem = openssl_pkey_get_private($private_key_path);
+
+	// Compute signature using the SHA512 signing algorithm
+	openssl_sign($message_digest, $signature, $private_key_pem, OPENSSL_ALGO_SHA512);
```

Comment:
return value not checked

Reply:
The return value is '$signature' which is checked via:

```
// Check that the signature exists
if ($signature) {
    // The signature was successfully created.
} else {
    //The signature wasn't properly generated.
    error_page(tra("Unable to produce openssl signature. Please try again at a later time."));
}
```

Do you believe this is an insufficient check?

Reply:
`openssl_sign` return boolean, which is not saved to variable, thus can't be checked.

Reply:
I've changed it to check this fully now, thanks for pointing this out. 👍